### ISSUE TITLE
[feature request] ofPolyline addVertex accepting glm::vec2

### ISSUE THREAD

**Issue by dimitre** [openframeworks/openFrameworks#7831]
it should just work, as ofPolyline is usually a 2d drawing.
some related discussion here: 
- https://github.com/openframeworks/openFrameworks/issues/5440

**Comment by artificiel** [comment]
definitely. I've skimmed through #5440 and TLDR is that ofPolyline is templated and things are not simple because T may == glm::vec2. this branch has work to support compile-time enabling by checking the types:

https://github.com/arturoc/openFrameworks/blob/1a2c94bfa52abf4f1f42ce934e0249746332a1f4/libs/openFrameworks/graphics/ofPolyline.h#L65-L69

then enabling the correct interpretation on the different glm::vec2 inputs such as:

https://github.com/arturoc/openFrameworks/blob/1a2c94bfa52abf4f1f42ce934e0249746332a1f4/libs/openFrameworks/graphics/ofPolyline.h#L90C1-L94C1

(but the branch has many other things going; perhaps the relevant bits can be pulled?)

**Comment by dimitre** [comment]
a new pr with relevant changes would be great!

**Comment by dimitre** [comment]
I've stumbled in this again and there are some things in the way of having a ```ofPolyline_<glm::vec2>```

this initialization
```c++
	void setRightVector(T v = T(0, 0, -1));
```

and this doesn't work
```c++
ofGetCurrentRenderer()->draw(*this);
```
because the template is already initialized with ofDefaultVertexType
```c++
using ofPolyline = ofPolyline_<ofDefaultVertexType>;
using ofDefaultVertexType = ofDefaultVec3;
```

**Comment by artificiel** [comment]
yes as mentionned above #5739 addresses all these problems but the work is not "done" — the PR is actually closed unfinished and it's scope is wider than 2d vs 3d (or maybe solving that pulls other things in). not clear if extracting the relevant bits is more productive than just redoing it (`if constexpr` seems more fluid for most of the cases, which is C++17, which was not allowed in OF at the moment of that PR. also `_v`).

**Comment by ofTheo** [comment]
I'd love easier glm::vec3 to glm::vec2 conversion in general!!! 

It is often a hangup with functions arguments and it often triggers really hard to decipher error messages. 
I'm not sure if there is a way to make a helper class or something to do this or a preprocessor macro, but I am guessing not. 

Wish there was a glm define to allow for automatic downgrading of glm::vec3 to vec2 and vice versa.

**Comment by artificiel** [comment]
the problem with ofPolyLine is not strictly conversion per se but that the code design of the class makes assumptions that the data is 3D. it should currently be named `of3DPolyLine` where the template stuff is private implementation. making it more flexible is what #5739 is after, where it can publicy become `ofPolyLine<glm::vec3>` or 2.

otherwise glm::vec3 -> 2 already works; 2 -> 3 does not because there is no unique correct way to default the value for 3rd member (often 0 but no necessarily). the old `ofVec3f` did it by assuming 0 for missing z, which is fine for 3D position in a (-1,1) space, but glm::vec3 gets used for other things too... hence glm deciding not supporting that conversion/construction.

https://github.com/openframeworks/openFrameworks/blob/24d166e077ecd8e2f01adf71fafad9db1ab11f15/libs/openFrameworks/math/ofVec3f.h#L111-L112

we could patch `glm/detail/type_vec2.hpp` for opt-in support:
```c++
#ifdef GLM_ALLOW_IMPLICIT_VEC2_TO_VEC3_CONVERSION_DEFAULTING_Z_TO_ZERO
    operator glm::vec3() const {
        return glm::vec3(x, y, 0.0f); 
    }
#endif
```
but while that might help, in itself it will not solve `ofPolyLine<glm::vec2>`.

also, if we focus on this specific issue as defined by it's title it's more of an user interface issue: it's simple to overload `ofPolyLine::addVertex(glm::vec2 vec)` and set the zero in the method (`ofPolyLine::addVertex( float x, float y, float z=0 )` already does something similar). there is a big diff in allowing ofPolyLine to ingest 2D, and having the whole structure optimized in 2D.

**Comment by dimitre** [comment]
@ofTheo I'm going in the opposite direction :D 
by removing implicit conversion and enforcing correct data types I think it is better to know what is going on (at least on OF Core) and sometimes a lot of implicit conversion is going on just to accept some int or double multiplication on GLM. like

```
vert = (vertices[i]+vertices[i+1]+vertices[i+2]) / 3;
```

it is probably converting glm::vec3 to ofVec3f to be able to multiply by int.

**Comment by artificiel** [comment]
@dimitre if you don't know "what is going on" FYI it has nothing to do with implicit conversion — glm::vec3 implements operator + and * and / on scalars (and a large number of other operators) so I'm not sure where the idea that "probably converting glm::vec3 to ofVec3f" comes from.

that being said I don't understand why implicit conversion is problematic. glm supports implicit conversion from glm::vec3 to ::vec2 — are you suggesting to remove that from glm?

**Comment by dimitre** [comment]
@artificiel I actually found this issues while removing implicit conversion, it is not from float, it is from double and int. 
just changing / 3 to /3.0f fixes that.

and glm is more strict with types conversion. 
OF is in fact extending glm to allow some more interoperability there

with pure glm it is not possible.
```
No viable conversion from 'vec<2, [2 * ...]>' to 'vec<3, [2 * ...]>'
```

**Comment by artificiel** [comment]
@dimitre i am getting more confused 

1. can you post a complete compilable snippet about your /3 vs /3.0 case?

2. what do you mean by "removing implicit conversion"?

3. "OF is in fact extending glm to allow some more interoperability there" — can you link to the source that extends glm?

4. as it is stated above that glm does not implement vec2 -> 3 because there is no clear default that makes sense. however it implements 3->2 because slicing is evident:
```c++
	glm::vec3 v3 {1,1,1};
	glm::vec2 v2 = v3;
	glm::vec3 v3b = v2;
```

**Comment by dimitre** [comment]
@artificiel yes without legacy OF math (ofVecXf) this is not possible. I've just tested your code and here is the result:
```
src/ofApp.cpp:7:12: error: no viable conversion from 'vec<2, [2 * ...]>' to 'vec<3, [2 * ...]>'
    7 |         glm::vec3 v3b = v2;
      |                   ^     ~~
../../../libs/macos/include/glm/./ext/../detail/type_vec3.hpp:108:17: note: candidate constructor not viable: no known conversion from 'glm::vec2' (aka 'vec<2, float, defaultp>') to 'const vec<3, float, (qualifier)0U> &' for 1st argument
  108 |                 GLM_CTOR_DECL vec(vec<3, T, Q> const& v) = default;
      |                               ^   ~~~~~~~~~~~~~~~~~~~~~
```

not many people realize, but the interop between vec2 and vec3 works in OF via
```
vec2 -> ofVec2f -> ofVec3f -> vec3
```

**Comment by artificiel** [comment]
@dimitre not sure i understand, as it's what i wrote above  was discussing glm, not ofVec stuff. 

>  4. (..) glm does not implement vec2 -> 3 because there is no clear default that makes sense. however it implements 3->2 because slicing is evident.

so `v3b = v2` does not compile. as you problably see in your test, `v2 = v3` does work. the implementation is 

https://github.com/g-truc/glm/blob/a583c59e1616a628b18195869767ea4d6faca5f4/glm/detail/type_vec2.hpp#L129-L130

the reason glm does not provide a conversion from 2 to 3 is not because conversion is not a good idea in itself, but that there is no consensus on the default of Z and it's preferable to express explicitely the intent than rely on a default.

**Comment by dimitre** [comment]
glm::vec2  has interoperability with ofVec2f, that can be promoted to ofVec3f and converted back to glm::vec3, under the hood
this is why nobody notices ofPolyline uses glm::vec3 until you work with list initializers or something more strict with types

**Comment by ofTheo** [comment]
Jumping back to the original issue - this works for me:


```
template<class T>
class ofPolyline_ {
public:
    using VertexType = T;
    
	/// \name Constructors
	/// \{

	/// \brief Creates an ofPolyline.
	ofPolyline_();

	/// \brief Creates an ofPolyline from a vector of glm::vec2 or T objects.
	ofPolyline_(const std::vector<T>& verts);

	static ofPolyline_ fromRectangle(const ofRectangle& rect);

    /// \}
    /// \name Add and Remove Vertices
    /// \{

	/// \brief Removes all the points from the ofPolyline.
	void clear();

	/// \brief Adds a point using an T at the end of the ofPolyline.
	void addVertex( const T& p );

    /// \brief Adds a point using floats at the end of the ofPolyline.
	void addVertex( float x, float y, float z=0 );
	
	void addVertex( glm::vec2 & vertex ){  //<------ new function 
		addVertex(vertex.x, vertex.y );
	}
```

I have to remove the forward declartion of glm::vec3 / glm::vec2 and include the glm headers for those two but it works. 

```
//#include <glm/detail/qualifier.hpp>
//namespace glm {
//	typedef vec<2, float, defaultp>		vec2;
//	typedef vec<3, float, defaultp>		vec3;
//}

#include <glm/vec2.hpp>
#include <glm/vec3.hpp>

```

**Comment by artificiel** [comment]
hey i guess i'm confused about topic drift... certainly `void addVertex( glm::vec2 & vertex )` is a solution to handle a "flat Z" polyline as an equivalence to `void addVertex( float x, float y, float z=0 )`; i was pointing to a branch where the polyline is wired differently for 2D vs 3D T's (currently makes assumptions about being 3D). 

discussion then drifted to conversions which i'm still not sure what the actual problem is? 

anyhow, it's trivial to add but seems very partial as there are other OF interfaces where glm::vec2 does not pass where x,y are expected; it would make sense to harmonize everywhere.
